File: hooks/inject-extension-rules-toml.py
#!/usr/bin/env python3
"""PreToolUse hook: inject config/*.toml based on file extension or git command."""

import json
import os
import re
import sys

CONFIG_DIR = os.path.join(os.path.expanduser("~"), ".claude", "config")


def read_config(filename: str) -> str:
    try:
        with open(os.path.join(CONFIG_DIR, filename)) as f:
            return f.read().strip()
    except OSError:
        return ""


def output(ctx: str) -> None:
    json.dump(
        {"hookSpecificOutput": {"hookEventName": "PreToolUse", "additionalContext": ctx}},
        sys.stdout,
    )


def main() -> None:
    data = json.load(sys.stdin)
    tool_input = data.get("tool_input", {})

    # Bash tool: git command -> inject git.toml
    command = tool_input.get("command", "")
    if command and re.search(r"(?:^|(?:&&|\|\||[;|`]|\$\()\s*)git\s", command):
        rules = read_config("git.toml")
        if rules:
            output(rules)
        return

    # Edit/Write tool: file extension -> inject {ext}.toml
    file_path = tool_input.get("file_path") or tool_input.get("path", "")
    if not file_path:
        return

    ext = os.path.splitext(file_path)[1].lstrip(".").lower()
    if not ext:
        return

    # 共通ルールを常に注入
    common = read_config("common.toml")

    rules = read_config(ext + ".toml")

    # .md -> also inject mmd.toml
    if ext == "md":
        mmd = read_config("mmd.toml")
        if mmd:
            rules = rules + "\n\n" + mmd if rules else mmd

    # common + extension-specific を結合
    if common:
        rules = common + "\n\n" + rules if rules else common

    if rules:
        output(rules)


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/inject_extension_rules_toml.rs
use serde_json::json;
use std::io::{self, Write};

use crate::input::HookInput;

fn config_dir() -> std::path::PathBuf {
    home::home_dir()
        .unwrap_or_default()
        .join(".claude")
        .join("config")
}

fn read_config(filename: &str) -> String {
    std::fs::read_to_string(config_dir().join(filename))
        .unwrap_or_default()
        .trim()
        .to_string()
}

fn output(ctx: &str) {
    let msg =
        json!({"hookSpecificOutput": {"hookEventName": "PreToolUse", "additionalContext": ctx}});
    let _ = writeln!(io::stdout(), "{msg}");
}

pub fn run(input: &HookInput) {
    let command = &input.tool_input.command;
    if !command.is_empty() {
        let re = regex::Regex::new(r"(?:^|(?:&&|\|\||[;|`]|\$\()\s*)git\s").unwrap();
        if re.is_match(command) {
            let rules = read_config("git.toml");
            if !rules.is_empty() {
                output(&rules);
            }
            return;
        }
    }

    let file_path = input.tool_input.file_path_resolved();
    if file_path.is_empty() {
        return;
    }

    let ext = std::path::Path::new(file_path)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("")
        .to_lowercase();
    if ext.is_empty() {
        return;
    }

    let common = read_config("common.toml");
    let mut rules = read_config(&format!("{ext}.toml"));

    if ext == "md" {
        let mmd = read_config("mmd.toml");
        if !mmd.is_empty() {
            rules = if rules.is_empty() {
                mmd
            } else {
                format!("{rules}\n\n{mmd}")
            };
        }
    }

    if !common.is_empty() {
        rules = if rules.is_empty() {
            common
        } else {
            format!("{common}\n\n{rules}")
        };
    }

    if !rules.is_empty() {
        output(&rules);
    }
}
